src/Waives.Pipelines/ConcurrentPipelineObserver.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Waives.Pipelines
+{
+    internal class ConcurrentPipelineObserver : IObserver<Document>, IDisposable
+    {
+        private readonly IDocumentProcessor _documentProcessor;
+        private readonly Action _onPipelineCompleted;
+        private bool _sourceComplete;
+
+        private readonly WorkPool _workPool;
+
+        internal ConcurrentPipelineObserver(
+            IDocumentProcessor documentProcessor,
+            Action onPipelineCompleted,
+            int maxConcurrency)
+        {
+            _documentProcessor = documentProcessor ?? throw new ArgumentNullException(nameof(documentProcessor));
+            _onPipelineCompleted = onPipelineCompleted ?? throw new ArgumentNullException(nameof(onPipelineCompleted));
+            _workPool = new WorkPool(maxConcurrency);
+        }
+
+        public void OnCompleted()
+        {
+            _sourceComplete = true;
+            _ = TryCompletionHandler();
+        }
+
+        public void OnError(Exception error)
+        {
+            throw new PipelineException(
+                $"A fatal error occurred in the processing pipeline: {error.Message}",
+                error);
+        }
+
+        public void OnNext(Document doc)
+        {
+            _workPool.Post(async () =>
+            {
+                try
+                {
+                    await _documentProcessor.RunAsync(doc);
+                }
+                finally
+                {
+                    await TryCompletionHandler();
```

Comment:
❓ Why do we need to signal possible pipeline completion for each document processed? We have the `OnCompleted()` handler for that.

Reply:
The `OnCompleted` method fires when the observer has received all source documents, but not when they've finished processing. So we need to check both in `OnCompleted` and in `OnNext`, because either can finish last. This is something that should go away with #62.